src/components/Toolbar.tsx

```diff
@@ -0,0 +1,371 @@
+  repoIconClass,
+  toolbarButtonClass,
+  toolbarClass,
+  toolbarMenuButtonClass,
+  toolbarMenuButtonIconClass,
+  toolbarMenuButtonSubtitleClass,
+  toolbarMenuButtonTitleClass,
+  toolbarMenuButtonTitleWrapperClass,
+  toolbarMenuWrapperClass,
+  toolbarNavClass
+} from '../style/Toolbar';
+import { GitCredentialsForm } from '../widgets/CredentialsBox';
+import { GitPullPushDialog, Operation } from '../widgets/gitPushPull';
+import { IGitExtension } from '../tokens';
+import { BranchMenu } from './BranchMenu';
+
+/**
+ * Displays an error dialog when a Git operation fails.
+ *
+ * @private
+ * @param model - Git extension model
+ * @param operation - Git operation name
+ * @returns Promise for displaying a dialog
+ */
+async function showGitOperationDialog(
+  model: IGitExtension,
+  operation: Operation
+): Promise<void> {
+  const title = `Git ${operation}`;
+  let result = await showDialog({
+    title: title,
+    body: new GitPullPushDialog(model, operation),
+    buttons: [Dialog.okButton({ label: 'DISMISS' })]
+  });
+  let retry = false;
+  while (!result.button.accept) {
+    retry = true;
+    const credentials = await showDialog({
+      title: 'Git credentials required',
+      body: new GitCredentialsForm(
+        'Enter credentials for remote repository',
+        retry ? 'Incorrect username or password.' : ''
+      ),
+      buttons: [Dialog.cancelButton(), Dialog.okButton({ label: 'OK' })]
+    });
+    if (!credentials.button.accept) {
+      break;
+    }
+    result = await showDialog({
+      title: title,
+      body: new GitPullPushDialog(model, operation, credentials.value),
+      buttons: [Dialog.okButton({ label: 'DISMISS' })]
+    });
+  }
+}
+
+/**
+ * Interface describing component properties.
+ */
+export interface IToolbarProps {
+  /**
+   * Git extension data model.
+   */
+  model: IGitExtension;
+
+  /**
+   * Boolean indicating whether branching is disabled.
+   */
+  branching: boolean;
+
+  /**
+   * Callback to invoke in order to refresh a repository.
+   *
+   * @returns promise which refreshes a repository
+   */
+  refresh: () => Promise<void>;
+}
+
+/**
+ * Interface describing component state.
+ */
+export interface IToolbarState {
+  /**
+   * Boolean indicating whether a branch menu is shown.
+   */
+  branchMenu: boolean;
+
+  /**
+   * Boolean indicating whether a repository menu is shown.
+   */
+  repoMenu: boolean;
+
+  /**
+   * Current repository.
+   */
+  repository: string;
+
+  /**
+   * Current branch name.
+   */
+  branch: string;
+}
+
+/**
+ * React component for rendering a panel toolbar.
+ */
+export class Toolbar extends React.Component<IToolbarProps, IToolbarState> {
+  /**
+   * Returns a React component for rendering a panel toolbar.
+   *
+   * @param props - component properties
+   * @returns React component
+   */
+  constructor(props: IToolbarProps) {
+    super(props);
+
+    const repo = this.props.model.pathRepository;
+    this._addListeners();
```

Comment:
Should be moved in `ComponentWillMount`

Reply:
Changed. Lifecycle method: `componentDidMount`.